### PR TITLE
Support ring-attention during sft phase

### PR DESCRIPTION
**Pull request by UbeCc**
This PR add support for ring-attention during sft phase. The implementation refers to dpo phase.

I have compared the results when w/  & wo/ ring attention. The curves are shown below.

image

The training scripts with ring-attn is: (dataset and model are open-sourced)
```
openrlhf.cli.train_sft \
   --max_len 512 \
   --dataset HKAIR-Lab/HK-O1aw-SFT-16K \
   --input_key prompt \
   --output_key answer \
   --train_batch_size 64 \
   --micro_train_batch_size 4 \
   --max_samples 30000 \
   --pretrain /mnt/nvme/models/Qwen2.5-Math-1.5B-Instruct/snapshots/aafeb0fc6f22cbf0eaeed126eff8be45b0360a35/ \
   --save_path ./checkpoint/test \
   --save_steps -1 \
   --logging_steps 1 \
   --eval_steps -1 \
   --zero_stage 3 \
   --max_epochs 2 \
   --bf16 \
   --flash_attn \
   --learning_rate 1e-5 \
   --load_checkpoint \
   --gradient_checkpointing \
   --packing_samples \
   --ring_attn_size 2 \
   --ring_head_stride 2
```

Thanks for your time and effort on reviewing the code!

### CODE REVIEW

File: openrlhf/datasets/sft_dataset.py [hunk 1]
```diff
@@ -158,12 +161,18 @@ def packing_collate_fn(self, item_list):
         index = 1
         for prompt_ids_len, input_id, attention_mask, info in item_list:
             packed_input_ids.append(input_id.flatten())
-            packed_attention_masks.append(torch.ones_like(input_id.flatten()) * index)
+            # packed_attention_masks.append(torch.ones_like(input_id.flatten()) * index)
```
Comment (zhuzilin) [inline openrlhf/datasets/sft_dataset.py:164]: please remove the commented line.

File: openrlhf/trainer/sft_trainer.py [hunk 2]
```diff
@@ -151,14 +163,20 @@ def fit(self, args, consumed_samples=0, num_update_steps_per_epoch=None):
...
-                        for label, source_len in zip(labels, prompts_id_lens):
+                        for label, source_len in zip(labels, prompt_id_lens):
                             label[:source_len] = self.loss_fn.IGNORE_INDEX
 
-                gpt_loss = self.loss_fn(output.logits, labels)
+                logits = output.logits                
+                if self.strategy.ring_attn_group is not None:
+                    total_seq_len = labels.numel()
+                    logits = all_gather(logits, self.strategy.ring_attn_group)
+                    logits = logits.reshape(output.logits.shape[0], total_seq_len, -1)
```
Comment (zhuzilin) [inline openrlhf/trainer/sft_trainer.py:177]: We'd better not all gather logits` as the logit is proportion to vocab_size. Could you initialize the loss_fn with the process group, and calculate logprob within, then do the allgather? In that way, we no longer need to write the code twice for both `train` and `evaluatue`.

File: openrlhf/trainer/sft_trainer.py [hunk 3]
```diff
@@ -270,4 +303,4 @@ def evaluate(self, eval_dataloader, steps=0):
                 elif self._tensorboard is not None:
                     for k, v in logs.items():
                         self._tensorboard.add_scalar(f"eval/{k}", v, steps)
-        self.model.train()  # reset model state
+        self.model.train()  # reset model state
```
Comment (zhuzilin) [inline openrlhf/trainer/sft_trainer.py:306]: please recover the origin newline.

**Comment by UbeCc** [comment]
image

I have change "all_gather on logits" to "all_reduce on loss". Here is the result

File: openrlhf/models/loss.py [hunk 4]
```diff
@@ -13,17 +13,32 @@ class GPTLMLoss(nn.Module):
     GPT Language Model Loss
     """
 
-    def __init__(self):
+    def __init__(self, process_group=None):
```
Comment (hijkzzz) [inline openrlhf/models/loss.py:16]: Please add some annotation about the process_group for Ring Attention (maybe rename to ring_process_group )
Thanks

Comment (UbeCc) [reply openrlhf/models/loss.py]: Yeah I have rename it to ring_attn_group in accordance with the name in strategy

**Comment by UbeCc** [comment]
image

I removed the dual-alignment process in `loss.py` because of redundancy
